sql/src/test/java/org/apache/druid/sql/calcite/DrillWindowQueryTest.java

```diff
@@ -142,65 +266,220 @@ public SpecificSegmentsQuerySegmentWalker createQuerySegmentWalker(
+    }
+
+    @Override
+    public void verify(String sql, List<Object[]> results)
+    {
+      List<Object[]> expectedResults = parseResults(currentRowSignature, expectedResultsText);
+      try {
+        Assert.assertEquals(StringUtils.format("result count: %s", sql), expectedResultsText.size(), results.size());
+        if (!isOrdered(sql)) {
+          results.sort(new ArrayRowCmp());
+          expectedResults.sort(new ArrayRowCmp());
+        } else {
+          assertResultsEquals(sql, expectedResults, results);
+        }
+      }
+      catch (AssertionError e) {
+        displayResults(expectedResults);
+        System.out.println("query: " + sql);
+        displayResults(results);
+        throw e;
+      }
+    }
+
+    private boolean isOrdered(String sql)
+    {
+      // FIXME: SqlToRelConverter.isOrdered(null) would be better
+      sql = sql.toLowerCase(Locale.ENGLISH).replace('\n', ' ');
+      sql = sql.substring(sql.lastIndexOf(')'));
+      return sql.contains("order");
+    }
+  }
+
+  static class ArrayRowCmp implements Comparator<Object[]>
+  {
+    @Override
+    public int compare(Object[] arg0, Object[] arg1)
+    {
+      String s0 = Arrays.toString(arg0);
+      String s1 = Arrays.toString(arg1);
+      return s0.compareTo(s1);
+    }
+  }
+
+  private static List<Object[]> parseResults(RowSignature rs, List<String[]> results)
+  {
+    Predicate<String> longPattern = Pattern.compile("^-?[0-9]+$").asPredicate();
+    Predicate<String> doublePattern = Pattern.compile("^-?[0-9]+\\.[0-9]+$").asPredicate();
+    List<Object[]> ret = new ArrayList<>();
+    for (String[] row : results) {
+      Object[] newRow = new Object[row.length];
+      List<String> cc = rs.getColumnNames();
+      for (int i = 0; i < cc.size(); i++) {
+        ColumnType type = rs.getColumnType(i).get();
+        assertNull(type.getComplexTypeName());
+        final String val = row[i];
+        Object newVal;
+        if ("null".equals(val)) {
+          newVal = null;
+        } else {
+          switch (type.getType()) {
+            case STRING:
+              newVal = val;
+              break;
+            case LONG:
+              if ("".equals(val)) {
+                newVal = null;
+              } else if (longPattern.test(val)) {
+                newVal = Numbers.parseLong(val);
+              } else if ("0E-20".equals(val)) {
+                newVal = 0L;
+              } else if (doublePattern.test(val)) {
+                double d = Doubles.tryParse(val);
+                newVal = (long) d;
+              } else {
+                Function<String, DateTime> parser = TimestampParser.createTimestampParser("auto");
+                try {
+                  newVal = parser.apply(val);
+                }
+                catch (IllegalArgumentException iae) {
+                  LocalTime v = LocalTime.parse(val);
+                  newVal = v.getMillisOfDay();
+                }
+              }
+              break;
```

Comment:
How about a flow such as below and avoid pattern matching? (0E-20 can be parsed using Doubles.tryParse)
if empty ; value = null
try Number.parseLong() 
  catch NFE
     try Doubes.parse
        catch NFE
           try timestampParsing